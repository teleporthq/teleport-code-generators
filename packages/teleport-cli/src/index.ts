#! /usr/bin/env node

import program from 'commander'
import chalk from 'chalk'
import minimist from 'minimist'
import clone from './commands/clone'
import watch from './commands/watch'
import format from './commands/format'
import sync from './commands/sync'
import init from './commands/init'
import packageJSON from '../package.json'
import { CONFIG_FILE } from './constants'

program.version(`v${packageJSON.version}`)

program.command('init').description(`Creates a ${CONFIG_FILE}`).action(init)

program
  .command('sync')
  .description(`Sync's all the components defined in ${CONFIG_FILE}`)
  .option('-f --force')
  .action(() => {
    const flags = minimist(process.argv.slice(2))
    sync({ force: Boolean(flags?.f || flags?.force) })
  })

program
  .command('format')
  .description(
    'Formats any file when a path is provided. Uses format options already set by @teleporthq'
  )
  .option('-p --path <path-url>')
  .action(() => {
    const flags = minimist(process.argv.slice(2))
    const targetPath = (flags?.p || flags?.path) ?? '/'
    format({ targetPath })
  })

program
  .command('watch')
  .description('Watch for a set of components in Studio and generate them in local')
  .option('-p --path <path-url>')
  .action(watch)

program
  .command('clone')
  .description('Pull a component from REPL / Studio')
  .option('-l --link')
  .option('-p --path <path-url>')
  .action(() => {
    const flags = minimist(process.argv.slice(2))
    if (Object.keys(flags).length === 0) {
      console.warn(chalk.yellow(`Link missing, please check --link`))
    }
    if (flags?.l || flags?.link) {
      clone({
        url: flags?.l || flags?.link,
        targetPath: flags?.p || flags?.path || '/',
      })
    }
  })

program.parse(process.argv)
