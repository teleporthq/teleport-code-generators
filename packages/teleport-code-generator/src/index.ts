import {
  PublisherResponse,
  ProjectUIDL,
  ComponentUIDL,
  Mapping,
  UIDLElement,
  GeneratorOptions,
} from '@teleporthq/teleport-types'
import { createProjectPacker } from '@teleporthq/teleport-project-packer'
import { ASSETS_IDENTIFIER } from '@teleporthq/teleport-shared/dist/cjs/constants'

import reactProjectGenerator, { ReactTemplate } from '@teleporthq/teleport-project-generator-react'
import nextProjectGenerator, {
  NextTemplate,
  NextMapping,
} from '@teleporthq/teleport-project-generator-next'
import vueProjectGenerator, { VueTemplate } from '@teleporthq/teleport-project-generator-vue'
import nuxtProjectGenerator, { NuxtTemplate } from '@teleporthq/teleport-project-generator-nuxt'
import preactProjectGenerator, {
  PreactTemplate,
} from '@teleporthq/teleport-project-generator-preact'
import stencilProjectGenerator, {
  StencilTemplate,
} from '@teleporthq/teleport-project-generator-stencil'

import { createZipPublisher } from '@teleporthq/teleport-publisher-zip'
import { createDiskPublisher } from '@teleporthq/teleport-publisher-disk'
import { createNowPublisher } from '@teleporthq/teleport-publisher-now'
import { createNetlifyPublisher } from '@teleporthq/teleport-publisher-netlify'
import { createGithubPublisher } from '@teleporthq/teleport-publisher-github'
import { createCodesandboxPublisher } from '@teleporthq/teleport-publisher-codesandbox'

import {
  createReactComponentGenerator,
  ReactStyleVariation,
} from '@teleporthq/teleport-component-generator-react'
import {
  createPreactComponentGenerator,
  PreactStyleVariation,
} from '@teleporthq/teleport-component-generator-preact'
import { createVueComponentGenerator } from '@teleporthq/teleport-component-generator-vue'
import { createStencilComponentGenerator } from '@teleporthq/teleport-component-generator-stencil'
import { createAngularComponentGenerator } from '@teleporthq/teleport-component-generator-angular'

import {
  PackerOptions,
  GenerateOptions,
  PublisherType,
  ProjectType,
  ComponentType,
  StyleVariation,
} from './types'

const projectGenerators = {
  [ProjectType.REACT]: reactProjectGenerator,
  [ProjectType.NEXT]: nextProjectGenerator,
  [ProjectType.VUE]: vueProjectGenerator,
  [ProjectType.NUXT]: nuxtProjectGenerator,
  [ProjectType.PREACT]: preactProjectGenerator,
  [ProjectType.STENCIL]: stencilProjectGenerator,
}

const templates = {
  [ProjectType.REACT]: ReactTemplate,
  [ProjectType.NEXT]: NextTemplate,
  [ProjectType.VUE]: VueTemplate,
  [ProjectType.NUXT]: NuxtTemplate,
  [ProjectType.PREACT]: PreactTemplate,
  [ProjectType.STENCIL]: StencilTemplate,
}

const projectPublisherFactories = {
  [PublisherType.ZIP]: createZipPublisher,
  [PublisherType.DISK]: createDiskPublisher,
  [PublisherType.NOW]: createNowPublisher,
  [PublisherType.NETLIFY]: createNetlifyPublisher,
  [PublisherType.GITHUB]: createGithubPublisher,
  [PublisherType.CODESANDBOX]: createCodesandboxPublisher,
}

const componentGeneratorFactories = {
  [ComponentType.REACT]: createReactComponentGenerator,
  [ComponentType.PREACT]: createPreactComponentGenerator,
  [ComponentType.ANGULAR]: createAngularComponentGenerator,
  [ComponentType.VUE]: createVueComponentGenerator,
  [ComponentType.STENCIL]: createStencilComponentGenerator,
}

const ComponentStyleVariations = {
  [ComponentType.REACT]: ReactStyleVariation,
  [ComponentType.PREACT]: PreactStyleVariation,
}

const reactStyledJSXGenerator = createReactComponentGenerator(ReactStyleVariation.StyledJSX, {
  mappings: [NextMapping as Mapping],
})

const createCodeGenerator = (factoryOptions: PackerOptions & GenerateOptions = {}) => {
  const {
    publisher = PublisherType.ZIP,
    projectType = ProjectType.NEXT,
    publishOptions = {},
    assets = [],
    componentType = ComponentType.REACT,
    styleVariation = ReactStyleVariation.CSSModules,
  } = factoryOptions

  const packer = createProjectPacker()

  const packProject = async (
    projectUIDL: ProjectUIDL,
    packOptions: PackerOptions = {}
  ): Promise<PublisherResponse<any>> => {
    const packProjectType = packOptions.projectType || projectType
    const packPublisher = packOptions.publisher || publisher
    const packPublishOptions = { ...publishOptions, ...packOptions.publishOptions }
    const packAssets = packOptions.assets || assets

    const projectGenerator =
      projectGenerators[packProjectType] || projectGenerators[ProjectType.NEXT]
    const projectTemplate = templates[packProjectType] || templates[ProjectType.NEXT]
    const publisherFactory =
      projectPublisherFactories[packPublisher] || projectPublisherFactories[PublisherType.ZIP]

    const projectPublisher = publisherFactory(packPublishOptions)

    packer.setAssets({
      assets: packAssets,
      path: [ASSETS_IDENTIFIER],
    })
    packer.setGenerator(projectGenerator)
    packer.setTemplate(projectTemplate)
    packer.setPublisher(projectPublisher)

    return packer.pack(projectUIDL)
  }

  const generateComponent = async (
    componentUIDL: ComponentUIDL,
    generateOptions: GenerateOptions = {}
  ) => {
    const generateComponentType = generateOptions.componentType || componentType
    const generateStyleVariation = generateOptions.styleVariation || styleVariation
    const generator = createComponentGenerator(generateComponentType, generateStyleVariation)

    return generator.generateComponent(componentUIDL)
  }

  const resolveElement = (node: UIDLElement, options?: GeneratorOptions) => {
    return reactStyledJSXGenerator.resolveElement(node, options)
  }

  return {
    packProject,
    generateComponent,
    resolveElement,
  }
}

export {
  createCodeGenerator,
  ProjectType,
  PublisherType,
  ComponentType,
  StyleVariation,
  ComponentStyleVariations,
  ReactStyleVariation,
  PreactStyleVariation,
  PackerOptions,
}

export default createCodeGenerator()

const createComponentGenerator = (componentType: string, styleVariation: StyleVariation) => {
  const generatorFactory = componentGeneratorFactories[componentType]

  if (componentType === ComponentType.REACT || componentType === ComponentType.PREACT) {
    return generatorFactory(styleVariation)
  }

  return generatorFactory()
}
