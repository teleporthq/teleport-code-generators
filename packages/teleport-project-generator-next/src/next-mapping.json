{
  "elements": {
    "navlink": {
      "elementType": "Link",
      "dependency": {
          "type": "library",
          "path": "next/link"
      },
      "attrs": {
        "href": { "type": "dynamic", "content": { "referenceType": "attr", "id": "transitionTo" } }
      },
      "children": [
        {
          "type": "element",
          "content": {
            "elementType": "a",
            "name": "link",
            "children": [
              { "type": "dynamic", "content": { "referenceType": "children", "id": "children" } }
            ]
          }
        }
      ]
    }
  }
}