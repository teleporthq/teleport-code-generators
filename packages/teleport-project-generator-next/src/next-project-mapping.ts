import { Mapping } from '@teleporthq/teleport-types'

export const NextProjectMapping: Mapping = {
  elements: {
    navlink: {
      elementType: 'Link',
      dependency: {
        type: 'library',
        path: 'next/link',
        version: '^12.1.0',
      },
      attrs: {
        href: { type: 'dynamic', content: { referenceType: 'attr', id: 'transitionTo' } },
      },
      children: [
        {
          type: 'element',
          content: {
            elementType: 'a',
            name: 'link',
            children: [{ type: 'dynamic', content: { referenceType: 'children', id: 'children' } }],
          },
        },
      ],
    },
    'html-node': {
      elementType: 'DangerousHTML',
      dependency: {
        type: 'package',
        path: 'dangerous-html',
        version: '0.1.12',
        meta: {
          importAlias: 'dangerous-html/react',
        },
      },
    },
    'lottie-node': {
      elementType: 'lottie-player',
    },
    fragment: {
      elementType: ' ',
      semanticType: '',
    },
  },
}
