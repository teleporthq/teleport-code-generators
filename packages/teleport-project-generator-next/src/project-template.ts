export default {
  name: 'teleport-project-next',
  files: [
    {
      name: 'package',
      content: `
{
  "name": "teleport-project-next",
  "version": "1.0.0",
  "description": "A next project generated based on a UIDL document",
  "main": "index.js",
  "scripts": {
    "dev": "next",
    "build": "next build",
    "start": "next start",
    "export": "next export"
  },
  "author": "TeleportHQ",
  "license": "MIT",
  "dependencies": {
    "next": "^9.0.2",
    "react": "^16.8.6",
    "react-dom": "^16.8.6"
  }
}`,
      fileType: 'json',
    },
  ],
  subFolders: [],
}
