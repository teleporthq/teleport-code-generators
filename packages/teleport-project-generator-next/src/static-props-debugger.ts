import { ComponentPlugin, ComponentPluginFactory } from '@teleporthq/teleport-types'
import traverse from '@babel/traverse'
import generate from '@babel/generator'
import { parse } from '@babel/parser'
import * as types from '@babel/types'

export const createStaticPropslogsPlugin: ComponentPluginFactory<{}> = () => {
  const staticPropsLogsPlugin: ComponentPlugin = async (structure) => {
    const { chunks } = structure

    const staticPropsChunk = chunks.find((chunk) => chunk.name === 'getStaticProps')
    if (staticPropsChunk === undefined) {
      return structure
    }
    const ast = staticPropsChunk.content as types.Node
    if (ast.type !== 'ExportNamedDeclaration') {
      return structure
    }

    const code = generate(types.program([ast])).code
    const parsedAST = parse(code, { sourceType: 'module' })

    traverse(parsedAST, {
      ObjectProperty(objectAst) {
        if (objectAst.node.value.type === 'OptionalMemberExpression') {
          objectAst.node.value = types.logicalExpression(
            '??',
            objectAst.node.value,
            types.objectExpression([])
          )
        }
      },
      BlockStatement(blockAst) {
        if (blockAst.parent.type === 'CatchClause') {
          blockAst.node.body = [
            types.expressionStatement(
              types.callExpression(types.identifier('console.trace'), [
                types.stringLiteral('Error:'),
                types.identifier('error'),
              ])
            ),
            ...blockAst.node.body,
          ]
        }
      },
      VariableDeclarator(variableAst) {
        if (
          variableAst.node.init.type === 'AwaitExpression' &&
          variableAst.node.id.type === 'Identifier' &&
          variableAst.parentPath.parentPath.node.type === 'BlockStatement'
        ) {
          const returnStatementIndex = variableAst.parentPath.parentPath.node.body.findIndex(
            (node) => node.type === 'ReturnStatement'
          )
          variableAst.parentPath.parentPath.node.body.splice(
            returnStatementIndex,
            0,
            types.expressionStatement(
              types.callExpression(types.identifier('console.log'), [
                types.identifier(variableAst.node.id.name),
              ])
            )
          )
        }
      },
    })

    structure.chunks = [
      ...chunks.filter((chunk) => chunk.name !== 'getStaticProps'),
      {
        ...staticPropsChunk,
        content: parsedAST.program.body[0],
      },
    ]

    return structure
  }

  return staticPropsLogsPlugin
}
