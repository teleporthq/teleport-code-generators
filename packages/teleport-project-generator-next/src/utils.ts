import {
  addAttributeToJSXTag,
  addChildJSXTag,
} from '@teleporthq/teleport-shared/dist/cjs/utils/ast-jsx-utils'
import { createJSXTag } from '@teleporthq/teleport-shared/dist/cjs/builders/ast-builders'

import * as types from '@babel/types'

import { prefixAssetsPath } from '@teleporthq/teleport-shared/dist/cjs/utils/uidl-utils'
import { FILE_TYPE, CHUNK_TYPE } from '@teleporthq/teleport-shared/dist/cjs/constants'
import { ProjectUIDL, ChunkDefinition, EntryFileOptions } from '@teleporthq/teleport-types'

export const createDocumentFileChunks = (uidl: ProjectUIDL, options: EntryFileOptions) => {
  const { settings, meta, assets, manifest } = uidl.globals

  const htmlNode = createJSXTag('html')
  const headNode = createJSXTag('Head')
  const bodyNode = createJSXTag('body')

  const mainNode = createJSXTag('Main')
  const nextScriptNode = createJSXTag('NextScript')
  addChildJSXTag(bodyNode, mainNode)
  addChildJSXTag(bodyNode, nextScriptNode)

  addChildJSXTag(htmlNode, headNode)
  addChildJSXTag(htmlNode, bodyNode)

  if (settings.language) {
    addAttributeToJSXTag(htmlNode, 'lang', settings.language)
  }

  // NOTE: Title is added in per page, not in the layout file

  if (manifest) {
    const linkTag = createJSXTag('link')
    addAttributeToJSXTag(linkTag, 'rel', 'manifest')
    addAttributeToJSXTag(linkTag, 'href', `${options.assetsPrefix}/manifest.json`)
    addChildJSXTag(headNode, linkTag)
  }

  meta.forEach((metaItem) => {
    const metaTag = createJSXTag('meta')
    Object.keys(metaItem).forEach((key) => {
      const metaValue = prefixAssetsPath(options.assetsPrefix, metaItem[key])
      addAttributeToJSXTag(metaTag, key, metaValue)
    })
    addChildJSXTag(headNode, metaTag)
  })

  assets.forEach((asset) => {
    const assetPath = prefixAssetsPath(options.assetsPrefix, asset.path)

    // link canonical for SEO
    if (asset.type === 'canonical' && assetPath) {
      const linkTag = createJSXTag('link')
      addAttributeToJSXTag(linkTag, 'rel', 'canonical')
      addAttributeToJSXTag(linkTag, 'href', assetPath)
      addChildJSXTag(headNode, linkTag)
    }

    // link stylesheet (external css, font)
    if ((asset.type === 'style' || asset.type === 'font') && assetPath) {
      const linkTag = createJSXTag('link')
      addAttributeToJSXTag(linkTag, 'rel', 'stylesheet')
      addAttributeToJSXTag(linkTag, 'href', assetPath)
      addChildJSXTag(headNode, linkTag)
    }

    // inline style
    if (asset.type === 'style' && asset.content) {
      const styleTag = createJSXTag('style')
      addAttributeToJSXTag(styleTag, 'dangerouslySetInnerHTML', { __html: asset.content })
      addChildJSXTag(headNode, styleTag)
    }

    // script (external or inline)
    if (asset.type === 'script') {
      const scriptTag = createJSXTag('script')
      addAttributeToJSXTag(scriptTag, 'type', 'text/javascript')
      if (assetPath) {
        addAttributeToJSXTag(scriptTag, 'src', assetPath)
        if (asset.options && asset.options.defer) {
          addAttributeToJSXTag(scriptTag, 'defer', true)
        }
        if (asset.options && asset.options.async) {
          addAttributeToJSXTag(scriptTag, 'async', true)
        }
      } else if (asset.content) {
        addAttributeToJSXTag(scriptTag, 'dangerouslySetInnerHTML', { __html: asset.content })
      }

      if (asset.options && asset.options.target === 'body') {
        addChildJSXTag(bodyNode, scriptTag)
      } else {
        addChildJSXTag(headNode, scriptTag)
      }
    }

    // icon
    if (asset.type === 'icon' && assetPath) {
      const iconTag = createJSXTag('link')
      addAttributeToJSXTag(iconTag, 'rel', 'shortcut icon')
      addAttributeToJSXTag(iconTag, 'href', assetPath)

      if (asset.options && asset.options.iconType) {
        addAttributeToJSXTag(iconTag, 'type', asset.options.iconType)
      }
      if (asset.options && asset.options.iconSizes) {
        addAttributeToJSXTag(iconTag, 'sizes', asset.options.iconSizes)
      }

      addChildJSXTag(headNode, iconTag)
    }
  })

  // Create AST representation of the class CustomDocument extends Document
  // https://github.com/zeit/next.js#custom-document
  const fileAST = createDocumentWrapperAST(htmlNode)

  const chunks: Record<string, ChunkDefinition[]> = {
    [FILE_TYPE.JS]: [
      {
        name: 'document',
        type: CHUNK_TYPE.AST,
        fileType: FILE_TYPE.JS,
        content: fileAST,
        linkAfter: [],
      },
    ],
  }

  return chunks
}

const createDocumentWrapperAST = (htmlNode, t = types) => {
  return t.program([
    t.importDeclaration(
      [
        t.importDefaultSpecifier(t.identifier('Document')),
        t.importSpecifier(t.identifier('Head'), t.identifier('Head')),
        t.importSpecifier(t.identifier('Main'), t.identifier('Main')),
        t.importSpecifier(t.identifier('NextScript'), t.identifier('NextScript')),
      ],
      t.stringLiteral('next/document')
    ),
    t.classDeclaration(
      t.identifier('CustomDocument'),
      t.identifier('Document'),
      t.classBody([
        t.classMethod(
          'method',
          t.identifier('render'),
          [],
          t.blockStatement([t.returnStatement(htmlNode)])
        ),
      ])
    ),
    t.exportDefaultDeclaration(t.identifier('CustomDocument')),
  ])
}
