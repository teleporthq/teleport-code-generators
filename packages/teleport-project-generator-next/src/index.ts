import { createProjectGenerator } from '@teleporthq/teleport-project-generator'
import { createComponentGenerator } from '@teleporthq/teleport-component-generator'
import {
  createReactComponentGenerator,
  ReactStyleVariation,
} from '@teleporthq/teleport-component-generator-react'

import { createPlugin as createHeadConfigPlugin } from '@teleporthq/teleport-plugin-jsx-head-config'
import prettierJS from '@teleporthq/teleport-postprocessor-prettier-js'

import { Mapping } from '@teleporthq/teleport-types'

import { createDocumentFileChunks } from './utils'
import NextMapping from './next-mapping.json'
import NextTemplate from './project-template'

const headConfigPlugin = createHeadConfigPlugin({
  configTagIdentifier: 'Head',
  configTagDependencyPath: 'next/head',
})

const createNextProjectGenerator = () => {
  const reactComponentGenerator = createReactComponentGenerator(ReactStyleVariation.StyledJSX)
  reactComponentGenerator.addMapping(NextMapping as Mapping)

  const reactPageGenerator = createReactComponentGenerator(ReactStyleVariation.StyledJSX, {
    plugins: [headConfigPlugin],
    mappings: [NextMapping as Mapping],
  })

  const documentFileGenerator = createComponentGenerator()
  documentFileGenerator.addPostProcessor(prettierJS)

  const generator = createProjectGenerator({
    components: {
      generator: reactComponentGenerator,
      path: ['components'],
    },
    pages: {
      generator: reactPageGenerator,
      path: ['pages'],
      options: {
        usePathAsFileName: true,
        convertDefaultToIndex: true,
      },
    },
    entry: {
      generator: documentFileGenerator,
      path: ['pages'],
      fileName: '_document',
      chunkGenerationFunction: createDocumentFileChunks,
    },
    static: {
      prefix: '/static',
      path: ['static'],
    },
  })

  return generator
}

export { createNextProjectGenerator, NextMapping, NextTemplate }

export default createNextProjectGenerator()
