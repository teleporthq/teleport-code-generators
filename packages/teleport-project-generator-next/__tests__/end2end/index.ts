import { FileType } from '@teleporthq/teleport-types'
import uidlSample from '../../../../examples/test-samples/project-sample.json'
import invalidUidlSample from '../../../../examples/test-samples/project-invalid-sample.json'
import uidlSampleWithoutProjectStyleesButImports from './project-with-import-without-global-styles.json'
import uidlSampleWithProjectStyleSheet from '../../../../examples/test-samples/project-with-import-global-styles.json'
import uidlSampleWithJustTokens from '../../../../examples/test-samples/project-with-only-tokens.json'
import template from './template-definition.json'
import { createNextProjectGenerator } from '../../src'

describe('React Next Project Generator', () => {
  const generator = createNextProjectGenerator()

  it('runs without crashing and adding external imports to _app.js', async () => {
    const outputFolder = await generator.generateProject(
      uidlSampleWithoutProjectStyleesButImports,
      template
    )
    const assetsPath = generator.getAssetsPath()

    const publicFolder = outputFolder.subFolders.find((folder) => folder.name === 'pages')
    const appFile = publicFolder.files.find((file) => file.name === '_app')

    expect(assetsPath).toBeDefined()
    expect(outputFolder.name).toBe(template.name)
    expect(outputFolder.files[0].name).toBe('package')
    expect(appFile).toBeDefined()
    expect(appFile.content).toContain(`import "antd/dist/antd.css`)
    expect(appFile.content).not.toContain(`import './style.css'`)
  })

  it('runs without crashing and adding style sheet to _app.js file', async () => {
    const outputFolder = await generator.generateProject(uidlSampleWithProjectStyleSheet, template)
    const assetsPath = generator.getAssetsPath()

    const publicFolder = outputFolder.subFolders.find((folder) => folder.name === 'pages')
    const appFile = publicFolder.files.find((file) => file.name === '_app')

    expect(assetsPath).toBeDefined()
    expect(outputFolder.name).toBe(template.name)
    expect(outputFolder.files[0].name).toBe('package')
    expect(appFile).toBeDefined()
    expect(appFile.content).not.toContain(`import "antd/dist/antd.css`)
    expect(appFile.content).toContain(`import './style.css'`)
  })

  it('runs without crashing', async () => {
    const outputFolder = await generator.generateProject(uidlSample, template)
    const assetsPath = generator.getAssetsPath()

    expect(assetsPath).toBeDefined()
    expect(outputFolder.name).toBe(template.name)
    expect(outputFolder.files[0].name).toBe('package')

    const components = outputFolder.subFolders[0]
    const pages = outputFolder.subFolders[1]

    expect(components.files[0].name).toBe('one-component')
    expect(pages.files[0].name).toBe('index')
    expect(pages.files[1].name).toBe('about')
  })

  it('runs without crashing and adds import of style sheet in _app.js', async () => {
    const result = await generator.generateProject(uidlSampleWithJustTokens, template)

    const pagesFolder = result.subFolders.find((folder) => folder.name === 'pages')
    const styleSheet = pagesFolder.files.find(
      (file) => file.name === 'style' && file.fileType === FileType.CSS
    )
    const appFile = pagesFolder.files.find(
      (file) => file.name === '_app' && file.fileType === FileType.JS
    )

    expect(styleSheet).toBeDefined()
    expect(styleSheet.content).toContain(`--greys-500: #595959`)
    expect(appFile).toBeDefined()
    expect(appFile.content).toContain(`import './style.css'`)
  })

  it('throws error when invalid UIDL sample is used', async () => {
    const result = generator.generateProject(invalidUidlSample, template)

    await expect(result).rejects.toThrow(Error)
  })
})
