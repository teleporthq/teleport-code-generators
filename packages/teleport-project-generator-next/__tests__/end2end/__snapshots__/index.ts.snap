// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `React Next Project Generator creates a next project and generates the named-slot for passing components 1`
] = `
Object {
  "content": "import React, { useEffect } from 'react'
import Head from 'next/head'

import AppComponent from '../components/component'
import Hero from '../components/hero'
import Footer from '../components/footer'

const Home = (props) => {
  useEffect(() => import('@lottiefiles/lottie-player'), [])
  return (
    <>
      <div className=\\"home-container1\\">
        <Head>
          <title>New Project3</title>
          <meta property=\\"og:title\\" content=\\"New Project3\\" />
        </Head>
        <lottie-player
          src=\\"https://assets9.lottiefiles.com/datafiles/gUENLc1262ccKIO/data.json\\"
          autoplay
          className=\\"home-lottie-node\\"
        ></lottie-player>
        <AppComponent
          fields={{
            Title: 'Custom Name',
          }}
        ></AppComponent>
        <Hero
          namedSlot={
            <div className=\\"home-container2\\">
              This is amazing, because this is a named-slot
              <div className=\\"home-container3\\">
                And can be nested any number of levels
              </div>
            </div>
          }
        ></Hero>
        <Footer></Footer>
      </div>
      <style jsx>
        {\`
          .home-container1 {
            width: 100%;
            display: flex;
            overflow: auto;
            min-height: 100vh;
            align-items: center;
            flex-direction: column;
            justify-content: flex-start;
          }
          .home-lottie-node {
            width: 200px;
            height: 200px;
          }
          .home-container2 {
            padding: 20px;
            font-size: 24px;
            background-color: var(--dl-color-success-700);
          }
          .home-container3 {
            padding: 20px;
            margin-top: 20px;
            background-color: var(--dl-color-danger-700);
          }
          @media (max-width: 479px) {
            .home-container1 {
              color: blue;
            }
          }
        \`}
      </style>
    </>
  )
}

export default Home
",
  "fileType": "js",
  "name": "index",
}
`

exports[
  `React Next Project Generator creates a next project and generates the named-slot for passing components 2`
] = `
Object {
  "content": "import React from 'react'

import PropTypes from 'prop-types'

const Hero = (props) => {
  return (
    <>
      <div className=\\"hero-hero\\">
        <div className=\\"hero-container1\\">
          <div className=\\"hero-container2\\">
            {props.namedSlot ?? (
              <div className=\\"hero-container3\\">
                This is a default value for named-slot
              </div>
            )}
          </div>
          <span className=\\"hero-text1\\">
            <span>
              <span>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed non
                volutpat turpis.
                <span
                  dangerouslySetInnerHTML={{
                    __html: ' ',
                  }}
                />
              </span>
              <span>
                <span
                  dangerouslySetInnerHTML={{
                    __html: ' ',
                  }}
                />
              </span>
            </span>
            <span>
              <span>
                Mauris luctus rutrum mi ut rhoncus. Integer in dignissim tortor.
                <span
                  dangerouslySetInnerHTML={{
                    __html: ' ',
                  }}
                />
              </span>
              <span>
                <span
                  dangerouslySetInnerHTML={{
                    __html: ' ',
                  }}
                />
              </span>
            </span>
          </span>
          <div className=\\"hero-btn-group\\">
            <button className=\\"hero-button1 button primary-button\\">
              {props.button}
            </button>
            <button className=\\"hero-button2 button secondary-button\\">
              {props.button1}
            </button>
          </div>
        </div>
        <img alt={props.imageAlt} src={props.imageSrc} className=\\"hero-image\\" />
      </div>
      <style jsx>
        {\`
          .hero-hero {
            width: 100%;
            display: flex;
            padding: var(--dl-space-space-threeunits);
            max-width: var(--dl-size-size-maxwidth);
            min-height: 80vh;
            align-items: center;
            flex-direction: row;
            justify-content: space-between;
          }
          .hero-container1 {
            display: flex;
            margin-right: var(--dl-space-space-threeunits);
            padding-right: var(--dl-space-space-threeunits);
            flex-direction: column;
          }
          .hero-container2 {
            font-size: 3rem;
            max-width: 450px;
            font-family: Super Funky;
          }
          .hero-text1 {
            margin-top: var(--dl-space-space-twounits);
            margin-bottom: var(--dl-space-space-twounits);
          }
          .hero-btn-group {
            display: flex;
            align-items: center;
            flex-direction: row;
          }
          .hero-button1 {
            transition: 0.3s;
            padding-top: var(--dl-space-space-unit);
            padding-left: var(--dl-space-space-twounits);
            padding-right: var(--dl-space-space-twounits);
            padding-bottom: var(--dl-space-space-unit);
          }
          .hero-button1:hover {
            color: var(--dl-color-gray-black);
            background-color: var(--dl-color-gray-white);
          }
          .hero-button2 {
            transition: 0.3s;
            margin-left: var(--dl-space-space-unit);
            padding-top: var(--dl-space-space-unit);
            padding-left: var(--dl-space-space-twounits);
            padding-right: var(--dl-space-space-twounits);
            padding-bottom: var(--dl-space-space-unit);
          }
          .hero-button2:hover {
            color: var(--dl-color-gray-white);
            border-color: var(--dl-color-gray-black);
            background-color: var(--dl-color-gray-black);
          }
          .hero-image {
            width: 400px;
            object-fit: cover;
          }
          .hero-container3 {
            padding: 20px;
            background-color: #f0f0f0;
          }
          @media (max-width: 991px) {
            .hero-hero {
              flex-direction: column;
            }
            .hero-container1 {
              align-items: center;
              margin-right: 0px;
              margin-bottom: var(--dl-space-space-twounits);
              padding-right: 0px;
            }
            .hero-container2 {
              text-align: center;
            }
            .hero-text1 {
              text-align: center;
              padding-left: var(--dl-space-space-threeunits);
              padding-right: var(--dl-space-space-threeunits);
            }
          }
          @media (max-width: 767px) {
            .hero-hero {
              padding-left: var(--dl-space-space-twounits);
              padding-right: var(--dl-space-space-twounits);
            }
            .hero-text1 {
              padding-left: var(--dl-space-space-unit);
              padding-right: var(--dl-space-space-unit);
            }
            .hero-image {
              width: 80%;
            }
          }
          @media (max-width: 479px) {
            .hero-hero {
              padding-top: var(--dl-space-space-twounits);
              padding-left: var(--dl-space-space-unit);
              padding-right: var(--dl-space-space-unit);
              padding-bottom: var(--dl-space-space-twounits);
            }
            .hero-container1 {
              margin-bottom: var(--dl-space-space-unit);
            }
            .hero-btn-group {
              flex-direction: column;
            }
            .hero-button2 {
              margin-top: var(--dl-space-space-unit);
              margin-left: 0px;
            }
          }
        \`}
      </style>
    </>
  )
}

Hero.defaultProps = {
  imageSrc:
    'https://images.unsplash.com/photo-1525498128493-380d1990a112?ixid=Mnw5MTMyMXwwfDF8c2VhcmNofDI0fHxtaW5pbWFsaXNtJTIwZ3JlZW58ZW58MHx8fHwxNjI1ODQxMDcw&ixlib=rb-1.2.1&h=1200',
  imageAlt: 'image',
  namedSlot: undefined,
  button: 'Get Started',
  button1: 'Learn More',
}

Hero.propTypes = {
  imageSrc: PropTypes.string,
  imageAlt: PropTypes.string,
  namedSlot: PropTypes.element,
  button: PropTypes.string,
  button1: PropTypes.string,
}

export default Hero
",
  "fileType": "js",
  "name": "hero",
}
`
